# Функции. Домашнее задание

**Обязательные задачи**

*Задача 1. Конвертация валют*

Напишите функцию `rub_to_currency`, которая переводит деньги в рублях в одну из двух валют: USD или CNY. У функции должно быть два входных параметра:
1. `rub_amount` — деньги в рублях;
2.  `currency` — валюта, в которую нужно совершить перевод. Если второй параметр не указан, по умолчанию следует считать его равным USD.

На выходе функция должна вывести на экран, сколько рублей было дано в начале и сколько денег получилось в заданной валюте. В случае с китайским юанем также нужно вывести промежуточное значение в долларах. Если на вход в `currency` приходит невалидное значение, нужно вывести на экран сообщение: «Расчёт не может быть совершён».

Декомпозируйте данную задачу на несколько функций.

Проверьте, что функция `rub_to_currency` в последующих ячейках отработала верно.

Создайте вспомогательные функции для перевода в доллар и китайский юань
Считайте курс доллара к рублю следующим: 1USD = 70,99RUB, курс доллара к юаню: 1USD = 6,44CNY

In [129]:
# валюта форматированным текстом
def curency_to_str(currency_value, currency_type):
    return str(currency_value) + currency_type

# печатаем результат конвертации форматированным текстом
def print_conversion(source_value, source_type, result_value, result_type):
    print(f"было: {curency_to_str(source_value, source_type)}, стало: {curency_to_str(result_value, result_type)}")    
    
# конвертер rub->usd
def rub_tu_usd(rub_amount = 0):
    USD_RUB_RATE = 70.99
    return rub_amount / USD_RUB_RATE

# конвертер usd->cny
def usd_tu_cny(usd_amount = 0):
    CNY_USD_RATE = 6.44
    return usd_amount * CNY_USD_RATE

# многострочно печатаем результат конвертации рублей в заказанную валюту форматированным текстом
# функция без return
def rub_to_currency(rub_amount, currency = "USD"):
    if currency == "USD":
        usd_amount = rub_tu_usd(rub_amount)
        print_conversion(rub_amount, "RUB", usd_amount, "USD")
    elif currency == "CNY":
        usd_amount = rub_tu_usd(rub_amount)
        cny_amount = usd_tu_cny(usd_amount)
        print("Промежуточная конвертация в USD")
        print_conversion(rub_amount, "RUB", usd_amount, "USD")
        print_conversion(usd_amount, "USD", cny_amount, "CNY")
    else:
        print("Расчёт не может быть совершён")    

In [130]:
rub_to_currency(7200, 'USD')
rub_to_currency(7200, 'CNY')
rub_to_currency(7200, 'EUR')
rub_to_currency(7200)

было: 7200RUB, стало: 101.4227355965629USD
Промежуточная конвертация в USD
было: 7200RUB, стало: 101.4227355965629USD
было: 101.4227355965629USD, стало: 653.1624172418651CNY
Расчёт не может быть совершён
было: 7200RUB, стало: 101.4227355965629USD


### Примеры вывода

rub_to_currency(7200, 'USD')
было: 7200RUB, стало: 101.4227355965629USD

rub_to_currency(7200, 'CNY')
Конвертация в USD
было: 7200RUB, стало: 101.4227355965629USD
было: 101.4227355965629USD, стало: 653.1624172418651CNY

rub_to_currency(7200, 'EUR')
Расчёт не может быть совершён

rub_to_currency(7200)
было: 7200RUB, стало: 101.4227355965629USD

*Задача 2. Обработка анкет*

Вам выгрузили список с анкетами потенциальных клиентов, которые заинтересовались покупкой подписки на здоровую еду. Вам нужно обработать эти значения и сказать, какие анкеты подходят (а какие нет) для дальнейшего обзвона колл-центром.

Условия, которым должна удовлетворять анкета:
* Есть контактная информация: номер телефона.
* Контактный номер начинается либо с восьмёрки, либо с семёрки и состоит из десяти символов.
* Город проживания: Москва.

Пример анкеты выглядит следующим образом: `name=Аня,phone=8800234 , city=москва`. Заметьте, в значениях параметров городов, имён, номеров телефонов могут встретиться лишние пробелы. Попробуйте удалить их с помощью метода `replace`.


In [81]:
applications = [
         'name=Аня,phone=8800234 ,city=москва',
         'name=КОЛЯ,phone=8800900871 ,city=МОСКВА',
         'name=Валентина,phone=7950900871 ,city=волгоград',
         'name=,phone=7999901871,city=москва',
         'name=Иван,phone=7999901871,city=москва',
         'name=Инга,phone=,city=москва'

] 

In [131]:
#константы структурыы анкеты
NAME_POS = 0
PHONE_POS = 1
CITY_POS = 2

CITY_TEMPLATE = "москва"
PHONE_LENGTH = 10

# функция проверки валидности номера
def check_phone(phone):
    normalized_phone = phone.replace(" ", "")
    if len(normalized_phone) != PHONE_LENGTH:
        return False
    if normalized_phone[0] != "8" and normalized_phone[0] != "7":
        return False
    return True

# функция проверки валидности города
def check_city(city):
    normalized_city = city.lower()
    index = normalized_city.find(CITY_TEMPLATE)
    return index != -1

# функция проверки одной анкеты
# разбираем структуру вида 'name=Аня,phone=8800234 ,city=москва'
def check_application(application):
    data = application.split(",")
    if len(data) != 3:
        return False  # некорректная структура анкеты, вернем FALSE
    name = data[NAME_POS].split("=")[1]
    phone = data[PHONE_POS].split("=")[1]
    city = data[CITY_POS].split("=")[1]
    return check_phone(phone) and check_city(city)

# функция проверки списка анкет
def find_valid_applications(applications):
    for elem in applications:
        print(f"{elem} --- {check_application(elem)}")

In [132]:
find_valid_applications(applications)

name=Аня,phone=8800234 ,city=москва --- False
name=КОЛЯ,phone=8800900871 ,city=МОСКВА --- True
name=Валентина,phone=7950900871 ,city=волгоград --- False
name=,phone=7999901871,city=москва --- True
name=Иван,phone=7999901871,city=москва --- True
name=Инга,phone=,city=москва --- False


### Примеры вывода
find_valid_applications(applications)
name=Аня,phone=8800234 ,city=москва --- False
name=КОЛЯ,phone=8800900871 ,city=МОСКВА --- True
name=Валентина,phone=7950900871 ,city=волгоград --- False
name=,phone=7999901871,city=москва --- True
name=Иван,phone=7999901871,city=москва --- True
name=Инга,phone=,city=москва --- False

*Задача 3. Факториал*

Факториалом числа n называется произведение всех натуральных чисел от 1 до n.



> n! = 1 \* 2 \*...\* n

> 3! = 1 \* 2 \* 3 = 6

> 0! = 1

Напишите рекурсивную функцию нахождения факториала, где входным параметром является одно единственное число — n. Проверьте, что функция в последующих ячейках отработала верно.









In [133]:
# Ваш код здесь
def factorial(n):
    if n:
        return n * factorial(n-1)   
    else:
        return 1

In [134]:
factorial(1)

1

In [135]:
factorial(0)

1

In [136]:
factorial(4)

24

*Задача 4. Палиндром*

Палиндромом называется такое слово (или фраза), которое слева направо и справа налево читается одинаково. Нужно написать рекурсивную функцию, которая будет возвращать `True`, если слово палиндром, и `False`, если нет.

Будем считать, что пустая строка также является палиндромом.

Решение должно работать на строках без знаков препинания. Проверьте, что функция в последующих ячейках отработала верно.



In [139]:
# в ТЗ не оговорили, но в примерах требуется игнор регистра и пробелов
def normalized_text(text):
    new_text = text.replace(" ", "")
    new_text = new_text.lower()
    return new_text

# рекурсивно проверяем текст на палиндром
def palindrom_with_normalized_text(text):
    if len(text) < 2:
        return True  # пустая строка или центральный одиночный символ = палиндром
    else:
        return text[0] == text[-1] and palindrom_with_normalized_text(text[1:-1])
    
# уберем пробелы и регистры и проверяем на палиндром
def palindrom(text):
    text = normalized_text(text)
    return palindrom_with_normalized_text(text)        

In [140]:
palindrom('saippuakivikauppias')

True

In [141]:
palindrom('murmur')

False

In [142]:
palindrom('В резерв')

True

In [143]:
palindrom('Sator Arepo tenet opera rotas')

True

**Дополнительные задачи**

*Задача 5. Количество цифр в числе*

Напишите функцию, которая рассчитывает количество цифр в целом числе, приходящем ей на вход.

Рассмотрите два способа решения данной задачи: через строковые переменные и без использования строковых переменных.

**Пример:**
<br>
Input: `n = 1234567` 
<br>
Output: `7`

In [144]:
# будем искать число цифр, знак не интересует
def normalized_number(number):
    return abs(number)

# способ 1, конвертируем число в строку и смотрим ее длину
def len_number_by_str(number):
    return len(str(normalized_number(number)))

# способ 2, честно считаем разрядность в десятичной системе
def len_number_wo_str(number = 0):
    number = normalized_number(number)
    count = 1
    while (number > 9):
        count +=1
        number = number // 10
    return count

# печатем читабельный вид двух вариантов сравнения
def print_number_len(number):
    print(f"{number}: {len_number_by_str(number)}(1-ый способ), {len_number_wo_str(number)}(2-ой способ)")

In [145]:
print("1234567:", len_number_by_str(1234567))

print_number_len(1234567)
print_number_len(1)
print_number_len(0)
print_number_len(10)
print_number_len(9)
print_number_len(-34)
print_number_len(55555555555555555555)

1234567: 7
1234567: 7(1-ый способ), 7(2-ой способ)
1: 1(1-ый способ), 1(2-ой способ)
0: 1(1-ый способ), 1(2-ой способ)
10: 2(1-ый способ), 2(2-ой способ)
9: 1(1-ый способ), 1(2-ой способ)
-34: 2(1-ый способ), 2(2-ой способ)
55555555555555555555: 20(1-ый способ), 20(2-ой способ)


*Задача 6. Количество отличников и двоечников*

Дан список, в каждом элементе которого лежат оценки за финальный экзамен по русскому языку по одному классу. Найдите количество сдавших (оценка >= 3) и не сдавших (оценка < 3) экзамен в каждом классе, а также выведите результативность класса.

Результативность класса считается по следующей логике:
* если количество оценок 4 и 5 в классе > 95%, результативность отличная;
* если количество оценок 4 и 5 в классе < 30%, результативность плохая;
* в остальных случаях результативность средняя.

Декомпозируйте задачу с помощью функций. Добавьте возможность менять пороги результативности с помощью входных параметров.


In [146]:
scores = [
          [2, 4, 4, 5, 2, 3, 3, 1, 1],
          [3, 4, 2, 3, 3, 1, 5, 2, 4],
          [5, 3, 3, 4, 3, 2, 3, 5],
          [4, 2, 2, 4, 2],
          [4, 3, 3, 3, 3, 3, 3],
          [2, 1, 5, 3, 5, 5],
          [4, 4, 4, 4, 5, 5],
          [4, 1, 3, 4, 1, 1],

]

In [147]:
# константы типов оценок
GRADE_BAD = 0
GRADE_GOOD = 1
GRADE_GREAT = 2

# константы разделительных границ оценок классов
GREAT_DEFAULT_LIMIT = 0.95
GOOD_DEFAULT_LIMIT = 0.3

# константы позиций по сдавшим/провалившим
FAIL_EXAM = 0
PASS_EXAM = 1
CLASS_PERC = 2

# список частот по оценкам 
def get_grades_freq(scores):
    freq_list = [0]*5
    try:
        for elem in scores:
            freq_list[elem - 1] +=1
    except:
        print("Некорректная запись оценки", scores)
        return None
    return freq_list

# переводим в читабельный вид результирующую оценку класса
def class_summary(result_type):
    result_strs = ["Плохо", "Удовлетворительно", "Отлично"]
    return "Итог по классу:" + result_strs[result_type]

# по проценту сдавших определяем тип оценки класса
def result_type(result, limit_good, limit_great):
    if result > limit_great:
        return GRADE_GREAT
    elif result > limit_good:
        return GRADE_GOOD
    else:
        return GRADE_BAD
   
# считаем характеристические числа для класса
# сдавшие/несдавшие/процент "4"+"5"
def class_eigenvalues(scores):
    freq_list = get_grades_freq(scores)
    if not freq_list:
        return None
    perc = (freq_list[4] + freq_list[3]) / sum(freq_list)
    fail_ex = freq_list[0] + freq_list[1]                 # получили <3
    pass_ex = freq_list[2] + freq_list[3] + freq_list[4]  # получили >=3
    return [fail_ex, pass_ex, perc]

# анализиоуем оценки одного класса
# структуры вида [5, 3, 3, 4, 3, 2, 3, 5],
def get_class_info(scores, limit_good, limit_great):
    eigenvalues = class_eigenvalues(scores)
    if eigenvalues:
        summary = class_summary(result_type(eigenvalues[CLASS_PERC], limit_good, limit_great))
        print(f"Количество сдавших:{eigenvalues[PASS_EXAM]}, Количество проваливших:{eigenvalues[FAIL_EXAM]}, {summary}")
    else:
        print("Некорректные оценки для класса")

# печатаем форматированную информацию по списку классов
def get_info(scores, limit_good = GOOD_DEFAULT_LIMIT, limit_great = GREAT_DEFAULT_LIMIT):
    for elem in scores:
        get_class_info(elem, limit_good, limit_great)    

In [148]:
get_info(scores)

Количество сдавших:5, Количество проваливших:4, Итог по классу:Удовлетворительно
Количество сдавших:6, Количество проваливших:3, Итог по классу:Удовлетворительно
Количество сдавших:7, Количество проваливших:1, Итог по классу:Удовлетворительно
Количество сдавших:2, Количество проваливших:3, Итог по классу:Удовлетворительно
Количество сдавших:7, Количество проваливших:0, Итог по классу:Плохо
Количество сдавших:4, Количество проваливших:2, Итог по классу:Удовлетворительно
Количество сдавших:6, Количество проваливших:0, Итог по классу:Отлично
Количество сдавших:3, Количество проваливших:3, Итог по классу:Удовлетворительно


### Примеры вывода
Количество сдавших:5, Количество проваливших:4, Итог по классу:Удовлетворительно
Количество сдавших:6, Количество проваливших:3, Итог по классу:Удовлетворительно
Количество сдавших:7, Количество проваливших:1, Итог по классу:Удовлетворительно
Количество сдавших:2, Количество проваливших:3, Итог по классу:Удовлетворительно
Количество сдавших:7, Количество проваливших:0, Итог по классу:Плохо
Количество сдавших:4, Количество проваливших:2, Итог по классу:Удовлетворительно
Количество сдавших:6, Количество проваливших:0, Итог по классу:Отлично
Количество сдавших:3, Количество проваливших:3, Итог по классу:Удовлетворительно

In [149]:
# Добавьте возможность менять пороги результативности с помощью входных параметров.
try:
    limit_good = int(input("какой процент оценок выше 3 считаем удовлетворительным "))
    limit_great = int(input("какой процент оценок выше 3 считаем отличным "))
    get_info(scores, limit_good/100, limit_great/100)
except:
   print("Проблемы с вводом процентов, отработаем дефолтно")
   get_info(scores)

какой процент оценок выше 3 считаем удовлетворительным 30
какой процент оценок выше 3 считаем отличным 95
Количество сдавших:5, Количество проваливших:4, Итог по классу:Удовлетворительно
Количество сдавших:6, Количество проваливших:3, Итог по классу:Удовлетворительно
Количество сдавших:7, Количество проваливших:1, Итог по классу:Удовлетворительно
Количество сдавших:2, Количество проваливших:3, Итог по классу:Удовлетворительно
Количество сдавших:7, Количество проваливших:0, Итог по классу:Плохо
Количество сдавших:4, Количество проваливших:2, Итог по классу:Удовлетворительно
Количество сдавших:6, Количество проваливших:0, Итог по классу:Отлично
Количество сдавших:3, Количество проваливших:3, Итог по классу:Удовлетворительно


*Задача 7. Калькулятор*

На вход в ваш калькулятор из стандартного ввода приходит выражение, в котором есть только числа и операции сложения и умножения.

Нужно написать функцию, которая обрабатывает входную строку и выдаёт результат выражения на выходе.

Используйте декомпозицию с помощью функций. Проверьте, что последующие ячейки отработали верно.




In [150]:
# разберем элемент на множители и перемножим
def mult_expression(expression):
    mult_operands = expression.split("*")
    mult_value = 1
    for elem in mult_operands:
        mult_value *= float(elem)
    return mult_value

# разберем на слагаемые(неопределенной сложности) и просуммируем
def sum_expression(expression):
    sum_operands = expression.split("+")
    sum_value = 0
    for elem in sum_operands:
        sum_value += mult_expression(elem) # внутри слагаемого может быть несколько перемножений
    return sum_value
    
# формат выражения заточен на неотрицатиельные числа, ими и ограничимся
def extract_expression(expression):
    result = sum_expression(expression)
    print(round(result))   # по ТЗ просили числа, но по формату примеров вывод целочисленный, округлим


In [151]:
expression = '1+2*4*5+7*10*11'
extract_expression(expression)

811


In [152]:
extract_expression('1+2+3+10')

16


In [153]:
extract_expression('1*2*10')

20


### Примеры вывода
expression = '1+2*4*5+7*10*11'
extract_expression(expression)
811

extract_expression('1+2+3+10')
16

extract_expression('1*2*10')
20